client/src/main/java/io/cloudsoft/winrm4j/client/WinRmClient.java

```diff
@@ -408,18 +426,159 @@ private WinRm getService() {
         if (winrm != null) return winrm;
+    
+        RuntimeException lastException = null;
+        
+        try {
+            winrm = null;
+            doCreateServiceWithBean();
+            return winrm;
+        } catch (RuntimeException e) {
+            LOG.warn("Error creating WinRm service with mbean strategy (trying other strategies): "+e, e);
+            lastException = e;
+        }
+        
+        /*
+         * It's tedious getting the right Provider esp in OSGi.
+         * 
+         * We've tried a bunch of strategies, with the most promising tried here,
+         * and detailed notes below.
+         */
+        
+        try {
+            winrm = null;
+            doCreateServiceWithReflectivelySetDelegate();
+            getAddressingProperties((BindingProvider) winrm);
+            return winrm;
+        } catch (RuntimeException e) {
+            LOG.warn("Error creating WinRm service with reflective delegate (trying other strategies): "+e, e);
+            lastException = e;
+        }
+        
+        try {
+            winrm = null;
+            doCreateServiceNormal();
+            return winrm;
+        } catch (RuntimeException e) {
+            LOG.warn("Error creating WinRm service with many strategies (giving up): "+e, e);
+            lastException = e;
+        }
+        
+        throw lastException;
+
+        // works, but addressing context might be null
+//        doCreateServiceWithReflectivelySetDelegate();
+        
+        // fails with NPE setting up feature (Bus is null)
+        // but it works if you install into karaf:  <feature>cxf-ws-addr</feature>
+//        doCreateServiceWithBean();
+        // also fails with NPE without that feature installed; untested with it
+//        doCreateServiceInSpecialClassLoader( ProviderImpl.class.getClassLoader() );
+//        doCreateServiceInSpecialClassLoader( JaxWsProxyFactoryBean.class.getClassLoader() );
+        
+        // fails in OSGi with CNF error when FactoryFinder tries to load the CXF impl 
+//        doCreateServiceWithSystemPropertySet();
+        
+        // fails in OSGi, with original error:
+        // com.sun.xml.internal.ws.client.sei.SEIStub cannot be cast to org.apache.cxf.frontend.ClientProxy
+        // at: ClientProxy.getClient(...);
+//        doCreateServiceNormal();
+//        doCreateServiceInSpecialClassLoader( WinRmClient.class.getClassLoader() );
+    }
+    
+    // normal approach
+    private synchronized void doCreateServiceNormal() {
+        WinRmService service = doCreateService_1_CreateMinimalServiceInstance();
+        Client client = doCreateService_2_GetClient(service);
+        doCreateService_3_InitializeClientAndService(client);
+    }
+
+    //  sys prop approach
+    private void doCreateServiceWithSystemPropertySet() {
+        System.setProperty("javax.xml.ws.spi.Provider", ProviderImpl.class.getName());
+        doCreateServiceNormal();
+    }
+    
+    // force delegate
```

Comment:
i think keep the comments and different strategies including the unused
private methods, for now -- if it works nicely we'll simplify it
dramatically but at least it will be in the history


On 31 August 2016 at 17:28, Вальо <notifications@github.com> wrote:

> In client/src/main/java/io/cloudsoft/winrm4j/client/WinRmClient.java
> <https://github.com/cloudsoft/winrm4j/pull/40#discussion_r77021768>:
>
> > +    }
> > +
> > +    // normal approach
> > +    private synchronized void doCreateServiceNormal() {
> > +        WinRmService service = doCreateService_1_CreateMinimalServiceInstance();
> > +        Client client = doCreateService_2_GetClient(service);
> > +        doCreateService_3_InitializeClientAndService(client);
> > +    }
> > +
> > +    //  sys prop approach
> > +    private void doCreateServiceWithSystemPropertySet() {
> > +        System.setProperty("javax.xml.ws.spi.Provider", ProviderImpl.class.getName());
> > +        doCreateServiceNormal();
> > +    }
> > +
> > +    // force delegate
>
> I prefer /** */ documenting.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cloudsoft/winrm4j/pull/40/files/f03798d6a81728efdb7bcf49495478cf5d84db37#r77021768>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAeTnAy9OoCkn2XWpaAFqtnXGeu7HhHaks5qlauqgaJpZM4JxzCm>
> .
>